### ISSUE TITLE
Snapshot downstairs still wanting to talk to SQLite

### ISSUE THREAD

**Issue by askfongjojo** [oxidecomputer/crucible#1377]
The issue was seen on rack2 after updating to `659130e0939fe27c4cabd38fad56734f89734407`:
```
root@oxz_switch0:~# pilot host exec -c 'svcs -xZ' 0-31
 6  BRM23230018        ok: 
 7  BRM27230045        ok: 
 8  BRM44220011        ok: svc:/oxide/crucible/downstairs:snapshot-fda723de-5177-4592-9963-f8f10fbd054e-e523f7a6-01db-4df5-bdab-a70e273be322 (Oxide Crucible Downstairs)
  Zone: oxz_crucible_0b41c560-3b20-42f4-82ad-92f5bb575d6b
 State: offline since Sat Jun 29 00:20:35 2024
Reason: Start method is running.
   See: http://illumos.org/msg/SMF-8000-C4
   See: /pool/ext/b93f880e-c55b-4d6c-9a16-939d84b628fc/crypt/zone/oxz_crucible_0b41c560-3b20-42f4-82ad-92f5bb575d6b/root/var/svc/log/oxide-crucible-downstairs:snapshot-fda723de-5177-4592-9963-f8f10fbd054e-e523f7a6-01db-4df5-bdab-a70e273be322.log
Impact: This service is not running.
 9  BRM44220005        ok: 
10  BRM42220009        ok: svc:/oxide/crucible/downstairs:snapshot-2d8b7eb9-32c9-48fb-88e7-197ac86dca8c-ea9ef2e0-8e90-451d-921a-97f1b625dd52 (Oxide Crucible Downstairs)
  Zone: oxz_crucible_cf3b2d54-5e36-4c93-b44f-8bf36ac98071
 State: offline since Sat Jun 29 00:20:38 2024
Reason: Start method is running.
   See: http://illumos.org/msg/SMF-8000-C4
   See: /pool/ext/d0584f4a-20ba-436d-a75b-7709e80deb79/crypt/zone/oxz_crucible_cf3b2d54-5e36-4c93-b44f-8bf36ac98071/root/var/svc/log/oxide-crucible-downstairs:snapshot-2d8b7eb9-32c9-48fb-88e7-197ac86dca8c-ea9ef2e0-8e90-451d-921a-97f1b625dd52.log
Impact: This service is not running.
11  BRM42220006        ok: 
12  BRM42220057        ok: 
14  BRM42220051        ok: svc:/oxide/crucible/downstairs:snapshot-9b7c581b-e432-486e-89ec-d9cfc9dfe792-250c960e-db9d-45df-9283-18231c4646b4 (Oxide Crucible Downstairs)
  Zone: oxz_crucible_ee44cdde-7ac9-4469-9f1d-e8bcfeb5cc46
 State: offline since Sat Jun 29 00:20:41 2024
Reason: Start method is running.
   See: http://illumos.org/msg/SMF-8000-C4
   See: /pool/ext/11b02ce7-7e50-486f-86c2-de8af9575a45/crypt/zone/oxz_crucible_ee44cdde-7ac9-4469-9f1d-e8bcfeb5cc46/root/var/svc/log/oxide-crucible-downstairs:snapshot-9b7c581b-e432-486e-89ec-d9cfc9dfe792-250c960e-db9d-45df-9283-18231c4646b4.log
Impact: This service is not running.
16  BRM42220014        ok: 
17  BRM42220017        ok: 
21  BRM42220031        ok: 
23  BRM42220016        ok: 
25  BRM44220010        ok: 
```

I looked at the log files in some of the crucible zones and saw errors like the one below:
```
root@oxz_crucible_800d1758:~# head -10 /var/svc/log/oxide-crucible-downstairs:snapshot-e9d4ed89-0862-4692-8b51-6ef39a90dac7-0b3e3b6a-c15b-44bd-86a8-c7abae144c72.log | looker  
[ Jun 29 00:00:02 Executing start method ("/opt/oxide/lib/svc/manifest/crucible/downstairs.sh"). ]
00:00:02.134Z INFO crucible: Opened existing region file "/data/regions/e9d4ed89-0862-4692-8b51-6ef39a90dac7/.zfs/snapshot/0b3e3b6a-c15b-44bd-86a8-c7abae144c72/region.json"
00:00:02.134Z INFO crucible: Database read version 1
00:00:02.134Z INFO crucible: Database write version 1
Error: SQLite extents are no longer supported
[ Jun 29 00:00:02 Stopping because service exited with an error. ]
```

**Comment by leftwo** [comment]
There are still snapshots with the old format:
```
root@oxz_crucible_2e32fdcc:~# more /data/regions/576099c1-6880-4908-ab9c-007b140cee03/.zfs/snapshot/7b314c0d-a185-4c17-b76f-482fd39f5c0c/region.json
{
  "block_size": 512,
  "extent_size": {
    "value": 131072,
    "shift": 9
  },
  "extent_count": 160,
  "uuid": "576099c1-6880-4908-ab9c-007b140cee03",
  "encrypted": true,
  "database_read_version": 1,
  "database_write_version": 1
}
```

And, our find did not find them I'm guessing:
```
oot@oxz_crucible_2e32fdcc:~# ls /data/regions/576099c1-6880-4908-ab9c-007b140cee03/.zfs/snapshot/7b314c0d-a185-4c17-b76f-482fd39f5c0c/00/000/000.db
/data/regions/576099c1-6880-4908-ab9c-007b140cee03/.zfs/snapshot/7b314c0d-a185-4c17-b76f-482fd39f5c0c/00/000/000.db
```

So, we either need to revert the sqlite backout, or, verify with an improved search to see if any old enough snapshots
exist on other racks

**Comment by askfongjojo** [comment]
It is likely that customer racks have some older snapshots. Here is a search of the sqlite files in the Colo rack:
```
root@oxz_switch1:~# pilot host exec -c 'ls /pool/ext/*/crypt/zone/oxz_crucible_*/root/data/regions/*/.zfs/snapshot/*/*/*/*.db | wc -l' 0-31
 0  BRM42220077        ok: 3920
 1  BRM44220012        ok: 128
 2  BRM42220001        ok: 1632
 3  BRM42220079        ok: 0
 4  BRM42220089        ok: 160
 5  BRM42220048        ok: 176
 6  BRM42220073        ok: 32
 7  BRM44220009        ok: 848
 8  BRM42220088        ok: 32
 9  BRM42220063        ok: 48
10  BRM44220017        ok: 0
11  BRM42220053        ok: 48
12  BRM42220028        ok: 48
13  BRM42220064        ok: 720
14  BRM42220013        ok: 176
15  BRM44220016        ok: 0
16  BRM42220078        ok: 80
17  BRM44220003        ok: 4480
18  BRM42220075        ok: 3968
19  BRM42220015        ok: 0
20  BRM42220035        ok: 112
21  BRM42220019        ok: 2112
22  BRM42220068        ok: 1440
23  BRM42220083        ok: 48
25  BRM42220032        ok: 368
26  BRM42220042        ok: 32
27  BRM44220021        ok: 96
28  BRM42220054        ok: 112
29  BRM42220087        ok: 848
30  BRM44220002        ok: 848
31  BRM42220022        ok: 32
```

**Comment by leftwo** [comment]
If you search for just a `000.db` file, you can get a count of the downstairs regions that
are still SQLite.  Here is that for dogfood:
```
root@oxz_switch0:~# pilot host exec -c 'ls /pool/ext/*/crypt/zone/oxz_crucible_*/root/data/regions/*/.zfs/snapshot/*/*/*/000.db | wc -l' 0-31
 6  BRM23230018        ok: 0
 7  BRM27230045        ok: 0
 8  BRM44220011        ok: 9
 9  BRM44220005        ok: 10
10  BRM42220009        ok: 13
11  BRM42220006        ok: 9
12  BRM42220057        ok: 12
14  BRM42220051        ok: 14
16  BRM42220014        ok: 11
17  BRM42220017        ok: 8
21  BRM42220031        ok: 11
23  BRM42220016        ok: 13
25  BRM44220010        ok: 7
```

This is a little different than the number of snapshots that exist:
```
root@oxz_switch0:~# pilot host exec -c 'zfs list -t snapshot | grep crucible | wc -l||true' 0-31
 6  BRM23230018        ok: 5
 7  BRM27230045        ok: 6
 8  BRM44220011        ok: 16
 9  BRM44220005        ok: 28
10  BRM42220009        ok: 33
11  BRM42220006        ok: 24
12  BRM42220057        ok: 30
14  BRM42220051        ok: 28
16  BRM42220014        ok: 23
17  BRM42220017        ok: 28
21  BRM42220031        ok: 23
23  BRM42220016        ok: 34
25  BRM44220010        ok: 21
```

Which, is also different than the number of downstairs that reported a failure when trying to start up
from the original report (outpuf from `pilot host exec -c 'svcs -xZ' 0-31`) which only showed three
downstairs that failed to start.

I wonder if we are also not cleaning up old snapshots?

**Comment by leftwo** [comment]
So, I did find that there are downstairs services that have started, but are failing and looping, and 
this might be missed by `svcs -xZ`.  For example, this services says it's online depending on when
you catch it, but the log shows the truth:
```
Error: SQLite extents are no longer supported                                                                                                                                  
[ Jul  1 16:31:42 Stopping because service exited with an error. ]                                                                                                             
[ Jul  1 16:31:42 Failing too quickly, throttling. ]                                                                                                                           
[ Jul  1 16:31:43 Executing start method ("/opt/oxide/lib/svc/manifest/crucible/downstairs.sh"). ]
{"msg":"Opened existing region file \"/data/regions/fda723de-5177-4592-9963-f8f10fbd054e/.zfs/snapshot/e523f7a6-01db-4df5-bdab-a70e273be322/region.json\"","v":0,"name":"crucib
le","level":30,"time":"2024-07-01T16:31:43.525363589Z","hostname":"oxz_crucible_0b41c560-3b20-42f4-82ad-92f5bb575d6b","pid":22423}
{"msg":"Database read version 1","v":0,"name":"crucible","level":30,"time":"2024-07-01T16:31:43.525608983Z","hostname":"oxz_crucible_0b41c560-3b20-42f4-82ad-92f5bb575d6b","pid
":22423}                                                                                                                                                                       
{"msg":"Database write version 1","v":0,"name":"crucible","level":30,"time":"2024-07-01T16:31:43.525620084Z","hostname":"oxz_crucible_0b41c560-3b20-42f4-82ad-92f5bb575d6b","pi
d":22423}                                                                                                                                                                      
Error: SQLite extents are no longer supported                                                                                                                                  
[ Jul  1 16:31:43 Stopping because service exited with an error. ]
[ Jul  1 16:31:43 Executing start method ("/opt/oxide/lib/svc/manifest/crucible/downstairs.sh"). ]                                                       
{"msg":"Opened existing region file \"/data/regions/fda723de-5177-4592-9963-f8f10fbd054e/.zfs/snapshot/e523f7a6-01db-4df5-bdab-a70e273be322/region.json\"","v":0,"name":"crucib
le","level":30,"time":"2024-07-01T16:31:43.612255616Z","hostname":"oxz_crucible_0b41c560-3b20-42f4-82ad-92f5bb575d6b","pid":22431}
{"msg":"Database read version 1","v":0,"name":"crucible","level":30,"time":"2024-07-01T16:31:43.612515106Z","hostname":"oxz_crucible_0b41c560-3b20-42f4-82ad-92f5bb575d6b","pid
":22431}         
{"msg":"Database write version 1","v":0,"name":"crucible","level":30,"time":"2024-07-01T16:31:43.61252737Z","hostname":"oxz_crucible_0b41c560-3b20-42f4-82ad-92f5bb575d6b","pid
":22431}            
Error: SQLite extents are no longer supported
[ Jul  1 16:31:43 Stopping because service exited with an error. ]                                                                                                             
[ Jul  1 16:31:43 Failing too quickly, throttling. ]
```

**Comment by leftwo** [comment]
Ok, so a little more makes sense here.

The number of snapshots can be different than the number of SQLite snapshots as newer snapshots (since build 5)
won't be SQLite as all volumes should have been converted to RAW. (unless you had a snapshot of an older snapshot..)

As far as downstairs services that were not reporting error, that can be explained by the check for `svcs -xZ` checking in 
the window of time that the service was running and had not yet decided it had an unsupported region and exiting.

If I look across the whole rack and find snapshot services that have restarted within the hour on todays date:

```
root@oxz_switch0:~# pilot host exec -c 'for zzz in $(svcs -Z -o fmri | grep snapshot); do echo -n "$zzz ";svcs -Zl $zzz | grep state_time; done | grep "1 19" | wc -l' 0-31
 6  BRM23230018        ok: 0
 7  BRM27230045        ok: 0
 8  BRM44220011        ok: 9
 9  BRM44220005        ok: 10
10  BRM42220009        ok: 13
11  BRM42220006        ok: 9
12  BRM42220057        ok: 12
14  BRM42220051        ok: 14
16  BRM42220014        ok: 11
17  BRM42220017        ok: 8
21  BRM42220031        ok: 11
23  BRM42220016        ok: 13
25  BRM44220010        ok: 7
```

Those numbers exactly match what `000.db` files we found.

**Comment by mkeeter** [comment]
This should be fixed in #1378
